eth/bor_checkpoint_verifier.go

```diff
@@ -1,61 +1,153 @@
+		return hash, errMissingBlocks
+	}
+
+	var localHash string
+
+	//verify the hash
+	if isCheckpoint {
+		var err error
+
+		//in case of checkpoint get the rootHash
+		localHash, err = handler.ethAPI.GetRootHash(ctx, start, end)
 
-		// verify the root hash of checkpoint
-		roothash, err := handler.ethAPI.GetRootHash(ctx, startBlock, endBlock)
 		if err != nil {
-			log.Debug("Failed to get root hash of checkpoint while whitelisting", "err", err)
-			return "", errRootHash
+			log.Debug("Failed to get root hash of given block range while whitelisting checkpoint", "start", start, "end", end, "err", err)
+			return hash, errRootHash
 		}
+	} else {
+		//in case of milestone(isCheckpoint==false) get the hash of endBlock
+		block, err := handler.ethAPI.GetBlockByNumber(ctx, rpc.BlockNumber(end), false)
+		if err != nil {
+			log.Debug("Failed to get end block hash while whitelisting milestone", "err", err)
+			return hash, errEndBlock
+		}
+
+		localHash = fmt.Sprintf("%v", block["hash"])[2:]
+	}
+
+	//nolint
+	if localHash != hash {
 
-		if roothash != checkpoint.RootHash.String()[2:] {
-			log.Warn("Checkpoint root hash mismatch while whitelisting", "expected", checkpoint.RootHash.String()[2:], "got", roothash)
-			return "", errCheckpointRootHashMismatch
+		log.Warn("Root hash mismatch while whitelisting", "expected", localHash, "got", hash)
+
+		ethHandler := (*ethHandler)(eth.handler)
+
+		var (
+			rewindTo uint64
+			doExist  bool
+		)
+
+		if doExist, rewindTo, _ = ethHandler.downloader.GetWhitelistedMilestone(); doExist {
+
+		} else if doExist, rewindTo, _ = ethHandler.downloader.GetWhitelistedCheckpoint(); doExist {
+
+		} else {
+			if start <= 0 {
+				rewindTo = 0
+			} else {
+				rewindTo = start - 1
+			}
 		}
 
-		// fetch the end checkpoint block hash
-		block, err := handler.ethAPI.GetBlockByNumber(ctx, rpc.BlockNumber(endBlock), false)
-		if err != nil {
-			log.Debug("Failed to get end block hash of checkpoint while whitelisting", "err", err)
-			return "", errEndBlock
+		if head-rewindTo > 255 {
+			rewindTo = head - 255
 		}
 
-		hash := fmt.Sprintf("%v", block["hash"])
+		rewindBack(eth, head, rewindTo)
+
+		return hash, errRootHashMismatch
+	}
+
+	// fetch the end block hash
+	block, err := handler.ethAPI.GetBlockByNumber(ctx, rpc.BlockNumber(end), false)
+	if err != nil {
+		log.Debug("Failed to get end block hash while whitelisting", "err", err)
+		return hash, errEndBlock
+	}
+
+	hash = fmt.Sprintf("%v", block["hash"])
+
+	return hash, nil
+}
+
+// Stop the miner if the mining process is running and rewind back the chain
+func rewindBack(eth *Ethereum, head uint64, rewindTo uint64) {
+	if eth.Miner().Mining() {
+		ch := make(chan struct{})
+		eth.Miner().Stop(ch)
+
+		<-ch
+		rewind(eth, head, rewindTo)
+
+		eth.Miner().Start(eth.etherbase)
+	} else {
+		rewind(eth, head, rewindTo)
+	}
+}
+
+func rewind(eth *Ethereum, head uint64, rewindTo uint64) {
+	log.Warn("Rewinding chain to :", rewindTo, "block number")
```

Comment:
This will throw an error because the order is interchanged. Let's make the log something like
```
log.Warn("Rewinding chain due to milestone whitelist mismatch", "to", rewindTo)
```
Also, please try to add some context to logs (like the reason of rewind) so that it becomes convenient and understandable for us as well as node operators.

Reply:
Okay!